app/services/district_report_service.rb

```diff
@@ -20,25 +23,52 @@ def call
     data
   end
 
-  def registrations_count(time)
-    Patient.where(registration_facility: @facilities).with_hypertension.where("recorded_at <= ?", time).count
-  end
-
   def compile_control_and_registration_data
-    @data[:cumulative_registrations] = registrations_count(selected_date.end_of_month)
-    -11.upto(0).each do |n|
+    months_of_data = [registration_counts.to_a.size, MAX_MONTHS_OF_DATA].min
+    @data[:cumulative_registrations] = lookup_registration_count(selected_date)
+    (-months_of_data + 1).upto(0).each do |n|
       time = selected_date.advance(months: n).end_of_month
       formatted_period = time.to_s(:month_year)
 
       @data[:controlled_patients][formatted_period] = controlled_patients(time).count
-      @data[:registrations][formatted_period] = registrations_count(time)
+      @data[:registrations][formatted_period] = lookup_registration_count(time)
     end
   end
 
   def format_quarter(quarter)
     "Q#{quarter.number}-#{quarter.year}"
   end
 
+  def lookup_registration_count(date)
+    lookup_date = date.beginning_of_month.to_date.to_s
+    row = registration_counts.find { |r| r["date"] == lookup_date }
+    return 0 unless row
+    row["running_ct"].to_i
+  end
+
+  def registration_counts
+    where_clause = ActiveRecord::Base.sanitize_sql_array([
+      "registration_facility_id in (?) and medical_histories.hypertension = ?",
+      @facilities, "yes"
+    ])
+
+    @registration_counts ||= Patient.connection.select_all(<<-SQL)
+      WITH cte AS (
+        SELECT date_trunc('month', "recorded_at") AS month, count(*) AS month_ct
```

Comment:
I think you'd potentially need to make `date_trunc` timezone sensitive here since we've stepped out of the rails world.

It's easy to show why `date_trunc` doesn't behave the way one would think it does by a few simple queries:

1. Regular `date_trunc`:

```
> select id, (DATE_TRUNC('month', recorded_at)) from blood_pressures limit 1;

                  id                  |     date_trunc
--------------------------------------+---------------------
 9cb6b949-f134-4abe-b0a4-47334c585b55 | 2019-11-01 00:00:00
```

2. `date_trunc` with time-zone resolution

```
> select id, (DATE_TRUNC('month', recorded_at)) AT TIME ZONE 'Asia/Kolkata' from blood_pressures limit 1;

                  id                  |         timezone
--------------------------------------+---------------------------
 9cb6b949-f134-4abe-b0a4-47334c585b55 | 2019-11-01 00:00:00+05:30

```

3. `date_trunc` with accurate timezone conversion

```
select id, (DATE_TRUNC('month', (recorded_at::timestamptz) AT TIME ZONE 'Asia/Kolkata')) AT TIME ZONE 'Asia/Kolkata' from blood_pressures limit 1;
                  id                  |         timezone
--------------------------------------+---------------------------
 9cb6b949-f134-4abe-b0a4-47334c585b55 | 2019-11-01 00:00:00+05:30
```

You can clearly see the difference in output between 1 & 2/3. `date_trunc` effectively kicks out the timezone by cutting up the date against UTC (default session tz) and returns a `timestamp without timezone`

2 & 3. are invisibly confusing but they mean different things.

2 means that it truncates the stored database value to a month (which is still in UTC) and then **assumes** the value in IST and converts it to timestamptz.

3 means that it **interprets** the value in the session timezone (UTC), convert it to IST, truncate it to a month and then convert it back to a timestamp with tz.


The subtle difference between 2. and 3. is that 2. assumes a UTC date to be in IST, 3. **corrects** the date to be in IST (just like Rails would). 1. doesn't do anything, it treats it in UTC. 

This is the idea behind `date_to_period_sql`.

Errors from this are usually seen when you hit month (or any period) boundaries.

Reply:
I _think_ I follow all of that. 😁 

I think for this initial ship for beta 1, it will be fine to let it all be done in UTC.  Converting everything in this query to correctly be in IST was proving to be really tricky, and I don't think its worth the complexity to just have a series of high level running totals be displayed in IST vs UTC.  These aren't 'per minute' values, so I don't think having the month cutoff be UTC will even be noticeable while we are in a beta early-access period.

Reply:
That seems fair.